### ISSUE TITLE
Custom expressions disappearing from the query builder

### ISSUE THREAD

**Issue by mazameli** [metabase/metabase#8215]
On our stats instance, questions that prevoiusly contained custom aggregation expressions in the query builder no longer show the expression, though the question still runs correctly:

![screen shot 2018-08-03 at 11 19 57 am](https://user-images.githubusercontent.com/2223916/43658951-434f5834-970f-11e8-86ed-8284a55b12fa.png)

If you click on `view the sql`, the custom expression still shows up in the generated SQL.

Clicking on `custom expression` in the Choose an Aggregation dropdown doesn't appear to do anything, but returns this JS error:

```
Uncaught metric with ID does not exist: 17
formatMetric @ formatter.js:70
format @ formatter.js:42
componentWillReceiveProps @ ExpressionEditorTextfield.jsx:82
componentWillMount @ ExpressionEditorTextfield.jsx:67
performInitialMount @ ReactCompositeComponent.js:348
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
mountChildren @ ReactMultiChild.js:234
_createInitialChildren @ ReactDOMComponent.js:701
mountComponent @ ReactDOMComponent.js:520
mountComponent @ ReactReconciler.js:43
mountChildren @ ReactMultiChild.js:234
_createInitialChildren @ ReactDOMComponent.js:701
mountComponent @ ReactDOMComponent.js:520
mountComponent @ ReactReconciler.js:43
_updateRenderedComponent @ ReactCompositeComponent.js:762
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
performUpdateIfNecessary @ ReactCompositeComponent.js:558
performUpdateIfNecessary @ ReactReconciler.js:154
runBatchedUpdates @ ReactUpdates.js:148
perform @ Transaction.js:141
perform @ Transaction.js:141
perform @ ReactUpdates.js:87
flushBatchedUpdates @ ReactUpdates.js:170
closeAll @ Transaction.js:207
perform @ Transaction.js:154
batchedUpdates @ ReactDefaultBatchingStrategy.js:60
batchedUpdates @ ReactUpdates.js:95
dispatchEvent @ ReactEventListener.js:145
```

Creating and saving a _new_ question using a custom expression seems to work as expected, even with the same data that's used in the other broken questions. So I'm wondering if this is a problem related to upgrading+migrating to 0.30.

**Comment by tlrobinson** [comment]
@mazameli @salsakran It looks like the metric used in the questions @mazameli pointed me to on stats have been archived (also matches the missing metric in the error above).

* `/question/191` has this `dataset_query`:
```
{
  "database":4,
  "type":"query",
  "query":{
    "source_table":9,
    "aggregation": [["METRIC",17]],
    "filter":["AND",["TIME_INTERVAL",80,-1,"day"]]
  }
}
```
* `/admin/datamodel/database/4/table/9` lists no metrics
* `/admin/datamodel/metric/17` exists but `archived = true`

Questions I have:

1. was this intentionally/unintentionally archived via user interaction, or accidentally archived via some migration or something. If the former I don't think we need to do anything for 0.30, if the latter then we definitely should.
2. the ux for questions that use archived metrics isn't great, but probably not something that needs to be addressed for 0.30

**Comment by mazameli** [comment]
Hmm, interesting. It highlights a glaring UX problem with the way we handle metric/segment lifecycle currently: there's no way in the UI to audit what happened to that metric, and no way to bring it back. Do we have any way of answering question #1 that you raised?

**Comment by salsakran** [comment]
this is probably out of scope for 0.30, but we should have a better way to deal with archived metrics.

This doesn't block 0.30 for me given that it has nothing to do with changes we made there

**Comment by mdz0r** [comment]
I'm having the same issue migrating from 0.29. Queries still work fine but I've lost my custom metrics. My metrics were not archived.

**Comment by mazameli** [comment]
@mdz0r just to clarify: did your queries use _Metrics_, which are defined in the Data Model of the Admin section, and which are referred to as "Common Metrics" in the query builder? Or were your queries using a _Custom Expression_ like this:

<img width="580" alt="screen shot 2018-08-09 at 9 59 44 am" src="https://user-images.githubusercontent.com/2223916/43913828-fdd8cabe-9bba-11e8-9acc-7763ae740d48.png">

**Comment by mdz0r** [comment]
Yes, metrics in the data model.

**Comment by flamber** [comment]
This was fixed by #9453

**Comment by tlrobinson** [comment]
@mazameli Did this issue actually have anything to do with *custom expressions*, or was it due to the missing *saved metrics*? If the latter then I agree with @flamber #9453 should resolve this.

**Comment by mazameli** [comment]
Saved Metrics. So yes, this is fixed by #9453.
